### PR TITLE
Update abuse-key for password recovery to add IP

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?
This PR adds IP Checking to the abuse-key for password recovery to help combat the ability to email harvest on Appwrite

## Related PRs and Issues

#2372 

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes,  I have

### CODE REVIEW

**Comment by eldadfux** [comment]
We need to have this as a separate check instead of adding the IP to the unique string, the string only defines what values would be considered as unique values for each request and will increase the abuse count. We need to apply a mechanism to support multiple abuse checks, each by a different unique key.

File: app/controllers/shared/api.php [hunk 1]
```diff
@@ -37,41 +37,60 @@
...
+    } else {
+        $timeLimit = new TimeLimit($abuseKeyLabel, $route->getLabel('abuse-limit', 0), $route->getLabel('abuse-time', 3600), $db);
+        $timeLimit->setNamespace('app_'.$project->getId());
+        $timeLimit
+            ->setParam('{userId}', $user->getId())
+            ->setParam('{userAgent}', $request->getUserAgent(''))
+            ->setParam('{ip}', $request->getIP())
+            ->setParam('{url}', $request->getHostname().$route->getPath())
+        ;
+        $timeLimitArray[] = $timeLimit;
```
Comment (TorstenDittmann) [inline app/controllers/shared/api.php:63]: To be more dry, you can just check if `$abuseKeyLabel` is not an array and push it to `$timeLimitArray` - that way you will only have one occurence of `new TimeLimit()`.

Something like:
```php
if (!is_array($abuseKeyLabel)) {
    $timeLimitArray[] = $abuseKeyLabel;
}
```

That way you can iterate over the array everytime 👍🏻

File: app/controllers/shared/api.php [hunk 2]
```diff
@@ -37,41 +37,60 @@
...
 
-    $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::$roles);
-    $isAppUser = Auth::isAppUser(Authorization::$roles);
+        if ($timeLimit->limit()) {
+            $response
+                ->addHeader('X-RateLimit-Limit', $timeLimit->limit())
+                ->addHeader('X-RateLimit-Remaining', $timeLimit->remaining())
+                ->addHeader('X-RateLimit-Reset', $timeLimit->time() + $route->getLabel('abuse-time', 3600))
+            ;
+        }
```
Comment (TorstenDittmann) [inline app/controllers/shared/api.php:83]: Looks like these headers will be set for the last item in `$timeLimit` 🤔

File: app/controllers/shared/api.php [hunk 3]
```diff
@@ -37,41 +37,60 @@
...
 
-    if (($abuse->check() // Route is rate-limited
-        && App::getEnv('_APP_OPTIONS_ABUSE', 'enabled') !== 'disabled') // Abuse is not diabled
-        && (!$isAppUser && !$isPrivilegedUser)) // User is not an admin or API key
-        {
-        throw new Exception('Too many requests', 429);
+        $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::$roles);
+        $isAppUser = Auth::isAppUser(Authorization::$roles);
```
Comment (TorstenDittmann) [inline app/controllers/shared/api.php:86]: This part is not necessary to be part of the loop.

**Review by TorstenDittmann — COMMENTED** [review]
Looking at this implementation we might need to be more smart.

Since the headers sent need to be the closest to the abuse. Meaning if the IP has only 2 requests left, but the email has 3 and comes last in the iteration - the headers are wrong.

We need to keep track what headers are the closest to be trigger the limit and set them after the loop.

File: app/controllers/shared/api.php (hunk 3 above)
Comment (PineappleIOnic) [reply app/controllers/shared/api.php:81]: Isn't this used to check if the user can bypass the restrictions?

Comment (PineappleIOnic) [reply app/controllers/shared/api.php:81]: Isn't this used to check if the user can bypass the restrictions?

File: app/controllers/shared/api.php [hunk 4]
```diff
@@ -37,41 +37,54 @@
...
+            ->setParam('{userId}', $user->getId())
+            ->setParam('{userAgent}', $request->getUserAgent(''))
+            ->setParam('{ip}', $request->getIP())
+            ->setParam('{url}', $request->getHostname().$route->getPath());
+        $timeLimitArray[] = $timeLimit;
     }
 
+    //TODO make sure we get array here
+
+    $closestLimit = 999;
```
Comment (TorstenDittmann) [inline app/controllers/shared/api.php:60]: Not really like the hardcoded `999`, lets make it `null` and add a new condition to the next lines:

```php
($timeLimit->remaining() < $closestLimit || is_null($closestLimit))
```

File: app/controllers/shared/api.php [hunk 5]
```diff
@@ -37,41 +37,54 @@
...
+    $abuseKeyLabel = $route->getLabel('abuse-key', 'url:{url},ip:{ip}');
+    $timeLimitArray = [];
 
-    foreach ($request->getParams() as $key => $value) { // Set request params as potential abuse keys
-        if(!empty($value)) {
-            $timeLimit->setParam('{param-'.$key.'}', (\is_array($value)) ? \json_encode($value) : $value);
-        }
+    if (!is_array($abuseKeyLabel)) {
+        $abuseKeyLabel = [$abuseKeyLabel];
     }
```
Comment (eldadfux) [inline app/controllers/shared/api.php:45]:
```suggestion
    $abuseKeyLabel =  (!is_array($abuseKeyLabel)) ? [$abuseKeyLabel] : $abuseKeyLabel;
```

Comment (TorstenDittmann) [reply app/controllers/shared/api.php:45]: This would be an assignment even if it's already an array 🤔

**Review by eldadfux — APPROVED** [review]

**Pull request merged**